Added "Introducing the new Okta Mobile SDKs" post (#1250)
oktadeveloper/okta-blog: _source/_posts/2022-08-30-introducing-the-new-okta-mobile-sdks.md

---
layout: blog_post
title: "Introducing the New Okta Mobile SDKs"
author: mike-nachbaur
by: internal-contributor
communities: [mobile]

> description: "Announcing Okta's new, modular, streamlined mobile SDKs! Add auth in a couple lines💥"

Reviewer: @havi, any ideas on how to make this sound better?

Author: "Announcing Okta's new Mobile SDKs: easier to use, way more features, and a fresh new lemon scent! 🍋"

Reviewer: ```suggestion
description: "Announcing Okta's new Mobile SDKs: easier to use, way more features, and a fresh new lemon scent! 🍋"
```

Just like within Okta OIDC, the [WebAuthenticationUI](https://okta.github.io/okta-mobile-swift/development/webauthenticationui/documentation/webauthenticationui) library is capable of using a configuration file named Okta.plist to define your client settings. From there, you can sign into your application using a single line of code.

```swift
// Sign in using the default configuration
let token = try await WebAuthentication.shared?.signIn(from: view.window)

// Save the user's token
let credential = try Credential.store(token)
```

The [shared property](https://okta.github.io/okta-mobile-swift/development/webauthenticationui/documentation/webauthenticationui/webauthentication/shared) simplifies how to create a [WebAuthentication](https://okta.github.io/okta-mobile-swift/development/webauthenticationui/documentation/webauthenticationui/webauthentication) instance, but if you want direct control of the configuration, you can simply create one in code.

```swift
let auth = WebAuthentication(

>     issuer: URL(string: "https://dev-133337.okta.com"),

Reviewer: @mikenachbaur-okta I had to change this from example.okta.com
`dev-133337` is a domain reserved for the blog, using something like `https://{yourOktaDomain}` would work too, your call 😄

Author: I usually use `example.okta.com`, how about that instead?

Reviewer: We actually have a lint rule against using `example.okta.com` 😰

try Credential.store(token, security: [
    .accessGroup("com.myapp.access-group")
])

// Secure the token with biometrics, and require
// the device to be unlocked
try Credential.store(token, security: [
    .accessibility(.unlocked),
    .accessControl(.biometryAny)
])
```

The tags option can be used to help manage the tokens stored in the keychain, which helps when your app works with multiple tokens or accounts. This is a feature I'm very excited about, due to the advanced use cases it unlocks. Let's dig into that in the next section.

#### Supporting multiple tokens / users

We now support storing multiple tokens or user accounts in a single application. This is a big improvement over the legacy Okta OIDC SDK, which made the assumption that an app would only ever have a single user at a time. Storing multiple tokens helps you create more sophisticated apps, such as:

* Multi-user applications that allow users to quick switch between profiles
* Applications that utilize app extensions, or widgets that wish to use separate tokens for each extension, for security reasons
* Apps that use separate tokens with different scopes for escalated access

> For a host of reasons, the AuthFoundation library now supports the storage and use of multiple tokens simultaneously, while still supporting traditional single-user applications via the Credential.default property.

Reviewer 2: ```suggestion
For a host of reasons, the AuthFoundation library now supports the storage and use of multiple tokens simultaneously, while still supporting traditional single-user applications via the [`Credential.default`]( https://okta.github.io/okta-mobile-swift/development/authfoundation/documentation/authfoundation/credential/default) property.
```
adding "code" tag here
also added the link to docs. thx!

Author: I think I missed that one. It should've been a link to the appropriate docs. https://okta.github.io/okta-mobile-swift/development/authfoundation/documentation/authfoundation/credential/default

Reviewer 2: Updated now, thanks for catching

}
```

Later, retrieving these values can be an equally straightforward operation.

```swift
if let secretData = try? Keychain.Search(account: "ssoSecret").get(),
   let secretToken = try? Keychain.Search(account: "ssoToken").get(),
   let ssoSecret = String(data: secretData, encoding: .utf8),
   let ssoToken = String(data: secretData, encoding: .utf8)
{
    // Exchange the tokens using TokenExchangeFlow
}
```

#### Outgoing request authorization

Ultimately, developers use the Okta Mobile SDK in order to securely access resources on a server somewhere. Your users authenticate using their credentials, and your application can use the resulting tokens to perform network requests on their behalf.

Since this is such a common operation, we decided to improve the developer experience by streamlining the process.

```swift
let url = URL(string: "https://api.example.com/notes")!
var request = URLRequest(url: url)
request.httpMethod = "POST"
request.httpBody = jsonBody

// Add appropriate HTTP authorization headers,
// automatically refreshing the token if needed.
await credential.authorize(&request)

let (data, response) = try await URLSession.shared
    .data(for: request)
```

> When using the Swift Concurrency [Credential.authorize()](https://okta.github.io/okta-mobile-swift/development/authfoundation/documentation/authfoundation/credential/authorize(_:)) async function, this will automatically refresh the credential's access token if it's expired, and add the appropriate HTTP authorization header to the request object. If you're not using Swift Concurrency, the same thing can be accomplished with the [refreshIfNeeded(graceInterval:completion:)](https://okta.github.io/okta-mobile-swift/development/authfoundation/documentation/authfoundation/credential/refreshifneeded(graceinterval:completion:)) and [authorize(request:)](https://okta.github.io/okta-mobile-swift/development/authfoundation/documentation/authfoundation/credential/authorize(request:)) functions.

Reviewer 2: ```suggestion
When using the Swift Concurrency [`Credential.authorize()`](https://okta.github.io/okta-mobile-swift/development/authfoundation/documentation/authfoundation/credential/authorize(_:)) async function, this will automatically refresh the credential's access token if it's expired, and add the appropriate HTTP authorization header to the request object. If you're not using Swift Concurrency, the same thing can be accomplished with the [`refreshIfNeeded(graceInterval:completion:)`](https://okta.github.io/okta-mobile-swift/development/authfoundation/documentation/authfoundation/credential/refreshifneeded(graceinterval:completion:)) and [`authorize(request:)`](https://okta.github.io/okta-mobile-swift/development/authfoundation/documentation/authfoundation/credential/authorize(request:)) functions.
```
added code tags 3 times
